src/uct/ib/base/ib_device.c

```diff
@@ -1049,3 +1077,87 @@ const char *uct_ib_wc_status_str(enum ibv_wc_status wc_status)
 {
     return ibv_wc_status_str(wc_status);
 }
+
+static ucs_status_t uct_ib_device_create_ah(struct ibv_ah_attr *ah_attr,
+                                            struct ibv_pd *pd,
+                                            struct ibv_ah **ah_p)
+{
+    char buf[128];
+    char *p, *endp;
+    struct ibv_ah *ah;
+
+    ah = ibv_create_ah(pd, ah_attr);
+    if (ah == NULL) {
+        p    = buf;
+        endp = buf + sizeof(buf);
+        snprintf(p, endp - p, "dlid=%d sl=%d port=%d src_path_bits=%d",
+                 ah_attr->dlid, ah_attr->sl,
+                 ah_attr->port_num, ah_attr->src_path_bits);
+        p += strlen(p);
+
+        if (ah_attr->is_global) {
+            snprintf(p, endp - p, " dgid=");
+            p += strlen(p);
+            inet_ntop(AF_INET6, &ah_attr->grh.dgid, p, endp - p);
+            p += strlen(p);
+            snprintf(p, endp - p, " sgid_index=%d traffic_class=%d",
+                     ah_attr->grh.sgid_index, ah_attr->grh.traffic_class);
+        }
+
+        ucs_error("ibv_create_ah(%s) failed: %m", buf);
+        return UCS_ERR_INVALID_ADDR;
+    }
+
+    *ah_p = ah;
+    return UCS_OK;
+}
+
+ucs_status_t uct_ib_device_create_ah_cached(uct_ib_device_t *dev,
+                                            struct ibv_ah_attr *ah_attr,
+                                            struct ibv_pd *pd,
+                                            struct ibv_ah **ah_p)
+{
+    uct_ib_ah_attr_t attr = {.attr = *ah_attr, .pd = pd};
+    khiter_t iter;
+    int ret;
+    ucs_status_t status;
+
+    ucs_spin_lock(&dev->ah_lock);
+
+    /* looking for existing AH with same attributes */
+    iter = kh_get(uct_ib_ah, &dev->ah_hash, attr);
+    if (iter == kh_end(&dev->ah_hash)) {
+        /* new AH */
+        status = uct_ib_device_create_ah(ah_attr, pd, ah_p);
+
+        if (status != UCS_OK) {
+            ucs_spin_unlock(&dev->ah_lock);
+            return status;
+        }
+
+        /* store AH in hash */
+        iter = kh_put(uct_ib_ah, &dev->ah_hash, attr, &ret);
+
+        /* failed to store - rollback */
+        if (iter == kh_end(&dev->ah_hash)) {
+            ibv_destroy_ah(*ah_p);
+            ucs_spin_unlock(&dev->ah_lock);
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        kh_value(&dev->ah_hash, iter) = *ah_p;
+    } else {
+        /* found existing AH */
+        *ah_p = kh_value(&dev->ah_hash, iter);
+    }
+
+    ucs_spin_unlock(&dev->ah_lock);
+
+    return UCS_OK;
+}
+
+void uct_ib_device_destroy_ah_cached(uct_ib_device_t *dev,
```

Comment:
maybe remove this function?

Reply:
this function is pair for create_ah_cached, for now it is not used, but in future we may add runtime configuration to disable AH cache, or logging...

I re-implemented this call as inline - it should not affect to anything.

do you think we should remove it?

Reply:
yes, if we need it in future let's add, the code is easier to read when there is no call to a function which does nothing

Reply:
ok, removed